src/org.xtuml.bp.ui.text/src/org/xtuml/bp/ui/text/activity/OpenDeclarationAction.java

```diff
@@ -0,0 +1,267 @@
+		} else {
+			throw new RuntimeException("Expecting Activity editor. Found: " + targetEditor.getClass().getName());
+		}
+
+		ITextSelection selection = (ITextSelection) editor.getSelectionProvider().getSelection();
+		IDocument doc = editor.getDocumentProvider().getDocument(editor.getEditorInput());
+
+		try {
+			int startLine = selection.getStartLine() + 1;
+			int startOffset = selection.getOffset();
+			IRegion findWord = editor.findWord(doc, startOffset);
+			startOffset = findWord.getOffset();
+			int lineOffset = doc.getLineOffset(startLine - 1);
+			int offsetWithinLine = startOffset - lineOffset + 1;
+			ActivityEditorInput input = (ActivityEditorInput) editor.getEditorInput();
+			NonRootModelElement modelElement = input.getModelElement();
+			Body_c bdy = OALPersistenceUtil.getOALModelElement(modelElement);
+		    boolean parsed = false;
+		    
+			if (bdy == null) {
+				partialParse(modelElement, offsetWithinLine, doc);
+				bdy = OALPersistenceUtil.getOALModelElement(modelElement);
+				if (bdy == null) {
+					// element not found, just return
+					return;
+				}
+				parsed = true;
+			}
+			NonRootModelElement declarationElement = (NonRootModelElement) bdy.Finddeclaration(offsetWithinLine,
+					startLine);
+			if (declarationElement == null && !parsed) {
+				// If the user edited the OAL, it may need to be parsed  
+				partialParse(modelElement, offsetWithinLine, doc);
+				declarationElement = (NonRootModelElement) bdy.Finddeclaration(offsetWithinLine,
+						startLine);
+			}
+			if (declarationElement != null) {
+				if (declarationElement instanceof VariableLocation_c) {
+					VariableLocation_c location = (VariableLocation_c) declarationElement;
+				    lineOffset = doc.getLineOffset(location.getLinenumber() - 1);
+					editor.selectAndReveal(lineOffset + location.getStartposition() - 1,
+							location.getEndposition() + 1 - location.getStartposition());
+				} else {
+					boolean showInME = shouldShowInME(declarationElement);
+					if (showInME) {
+						showInModelExplorer(declarationElement);
+					} else {
+						showInModelExplorer(declarationElement);
+						Package_c pkg = declarationElement.getFirstParentPackage();
+						IEditorPart editorPart = EditorUtil.openEditorForElement(pkg);
+						editorPart.getSite().getSelectionProvider()
+								.setSelection(new StructuredSelection(declarationElement));
+						zoomToSelected(editorPart);
+					}
+				}
+			}
+		} catch (BadLocationException | PartInitException e) {
+			TextPlugin.logError("Unable to locate line information for the text editor selection.", e);
+		}
+
+	}
+	
+	private void partialParse(NonRootModelElement modelElement, int offsetWithinLine, IDocument doc) {
+		// run a parse of this body
+		ParseRunnable parseRunner = new PartialParseRunnable(modelElement, doc.get(),
+				DocumentUtil.positionToLine(offsetWithinLine, doc),
+				DocumentUtil.positionToCol(offsetWithinLine, doc));
+		parseRunner.run();						
```

Comment:
Don't use the PartialParseRunnable. PartialParseRunnable is a subclass of ParseRunnable and all of the extra functionality it offers is disabled because it was not ready for promotion. Passing in line and column information causes the content assist calculations to be done and will add a good deal of time to the parse.

Reply:
in summary: use the normal ParseRunnable, and do not pass line/column information.

Reply:
used ParseRunnable